### PR TITLE
TW-3666(fix): multi-URL federation lookup with identity fallback and …

### PR DESCRIPTION
**Pull request by jbrnzl**
# Federated phonebook lookup: iterate all base_urls, optional direct identity lookup, unified merge

## Summary
This change aligns the mobile federated contact lookup flow with the product specification: every URL under m.federated_identity_services is used for the full request-token -> register -> hash-details -> lookup pipeline. Third_party_mappings are still resolvd per delegated server. When m.identity_server is configured and its URL has not appeared in any third_party_mappings response, a direct lookup is performed on that server. Federation, identity, and third-party results are merged per chunk.

## What changed

- FederationLookUpArgument now takes federationUrls (List<String>) and optional identityServerUrl instead of a single federationUrl to best suit the flow described here:
Capture d’écran 2026-04-21 à 16 34 50

- ContactsManager passes all federation base_urls from configuration plus the configured identity server URL into the interactor.
- FederationLookUpPhonebookContactInteractor builds one session per federation URL (and optionally for the identity server), recomputes hashes per server where hash_details may differ, runs lookupMxid for each federation session per chunk, tracks third-party server keys across chunks to skip redundant identity lookups, merges mapping sources via existing combineContacts, and preserves prior behavior where third-party sub-lookup failures do not fail the entire chunk.
- Tests extended for multi-URL merging, identity lookup when not delegated via third_party_mappings, and skipping identity when already delegated.

## Test recommendations
Run flutter test test/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor_test.dart.
On a device or simulator with contacts permission, sign in against a homeserver that exposes m.federated_identity_services (multiple base_urls if available) and optional m.identity_server; open the contacts flow and confirm lookups complete and merged contacts behave as before for a single-URL configuration.

## Demos
**Attach screenshots or videos demonstrating the changes**
No demos provided because no UI changes, only behavioral changes (more exploration)
If you think I should add some, feel free to ask.

## Summary by CodeRabbit

* **New Features**
  * Contact lookup now queries multiple federation servers and accepts optional identity server info.

* **Improvements**
  * Reduced redundant identity-server queries; normalized URLs (https/trailing slash) for consistent matching.
  * Chunked processing to preserve partial results and improve reliability on large lookups.

* **Tests**
  * Expanded coverage for multi-server federation lookups, identity-server behaviors, deduplication, and skipping redundant identity queries.

### CODE REVIEW

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 1]
```diff
@@ -321,14 +228,258 @@ class FederationLookUpPhonebookContactInteractor {
...
+        federationUrl: session.url,
+        request: FederationLookupMxidRequest(
+          addresses: hashing.hashToContactIdMappings.values
+              .expand((h) => h)
+              .toSet(),
+          algorithm: session.hashDetails.algorithms?.firstOrNull,
+          pepper: session.hashDetails.lookupPepper,
+        ),
+        registeredToken: session.registerToken.token!,
+      );
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: I agree 👍🏻

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 2]
```diff
@@ -321,14 +228,258 @@ class FederationLookUpPhonebookContactInteractor {
...
+        hashing: hashing,
+        target: fromMappings,
+      );
+
+      if (response.thirdPartyMappings != null &&
+          response.thirdPartyMappings!.isNotEmpty) {
+        globalSeenThirdPartyUrls.addAll(response.thirdPartyMappings!.keys);
+        try {
+          final thirdPartyContacts = await _handleThirdPartyMappings(
+            thirdPartyToHashes: response.thirdPartyMappings!,
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart:367]: Valid in identifying a potential format mismatch, but the proposed fix is not correct: convertToHttps is not a general URL normalizer: it is designed exclusively for raw domain strings (e.g. 'identity.example.com'). When applied to a full HTTPS URL such as 'https://identity.example.com', it splits on :, takes 'https' as the domain, and returns 'https://https/', which is semantically incorrect.

In production, identityServerUrl comes from Uri.toString() via the Matrix SDK and is always a full URL.

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: Addressed

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart:367]: Addressed

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 3]
```diff
@@ -321,14 +228,258 @@ class FederationLookUpPhonebookContactInteractor {
...
+        registeredToken: registerToken.token!,
+      );
+
+      if (hashDetails.lookupPepper?.isEmpty == true &&
+          hashDetails.algorithms?.isEmpty == true) {
+        firstFailureRef(
+          const GetHashDetailsFailure(exception: 'Hash details is empty'),
+        );
+        return null;
+      }
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart:478]: Addressed

File: test/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor_test.dart [hunk 4]
```diff
@@ -1437,6 +1437,299 @@ void main() {
...
+                    c.phoneNumbers?.any(
+                      (p) => p.matrixId == '@alice:matrix.org',
+                    ) ==
+                    true,
+              ),
+              'alice has matrixId from first server',
+              true,
+            ),
+          ]),
+        );
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply test/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor_test.dart]: Addressed

File: test/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor_test.dart [hunk 5]
```diff
@@ -1437,6 +1437,299 @@ void main() {
...
+          expect(
+            result,
+            emitsInOrder(<dynamic>[
+              const Right<Failure, Success>(GetPhonebookContactsLoading()),
+              isA<Right<Failure, Success>>(),
+            ]),
+          );
+          // Two lookups should have been made: one federation + one identity.
+          await Future.delayed(const Duration(milliseconds: 100));
+          expect(lookupCount, greaterThanOrEqualTo(2));
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply test/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor_test.dart]: Addressed

File: test/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor_test.dart [hunk 6]
```diff
@@ -1437,6 +1437,299 @@ void main() {
...
+              FederationIdentityLookupSuccess(newContacts: {}),
+            ),
+          );
+
+          final result = interactor.execute(argument: argumentWithIdentity);
+
+          await result.last;
+          // Only the federation lookup should have been called; the identity
+          // server setup lookup should have been skipped at chunk time.
+          expect(lookupCount, 1);
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply test/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor_test.dart]: Addressed

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 7]
```diff
@@ -321,14 +228,288 @@ class FederationLookUpPhonebookContactInteractor {
...
+    FederationTokenInformation? tokenInfo;
+    await _federationIdentityRequestTokenManager
+        .execute(federationTokenRequest: tokenRequest)
+        .then(
+          (state) => state.fold((failure) {}, (success) {
+            if (success is FederationIdentityRequestTokenSuccess) {
+              tokenInfo = success.tokenInformation;
+            }
+          }),
+        );
```
Comment (nqhhdev) [inline lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: Token request is not per-URL in _setupSession — it's called N times with the same tokenRequest (bug risk).

Do you have any ideas for improvement?

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 8]
```diff
@@ -321,14 +228,288 @@ class FederationLookUpPhonebookContactInteractor {
...
+            'third party mappings failed for ${session.url}',
+          );
+        }
+      }
+    }
+
+    if (sessions.isNotEmpty &&
+        successfulLookups == 0 &&
+        firstLookupError != null) {
+      throw firstLookupError;
```
Comment (nqhhdev) [inline lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: `firstLookupError` is typed Object?. Rethrowing an Object loses the original stack trace. Should be Error.throwWithStackTrace(firstLookupError!, firstStackTrace) or the catch block should capture and store the stack trace alongside the error

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 9]
```diff
@@ -405,6 +582,8 @@ class FederationLookUpPhonebookContactInteractor {
     return updatedContact;
   }
 
+  String _normalizeUrl(String url) => url.replaceFirst(RegExp(r'/+$'), '');
```
Comment (nqhhdev) [inline lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: `thirdPartyMappings` keys from the server sometimes come as plain hostnames (e.g. identity.example.com) without a scheme. If convertToHttps doesn't handle that case, _normalizeUrl won't match against a full https://identity.example.com URL, so the identity server would be queried twice. Can we add a test for this edge case, or document the assumption that keys always include the scheme?

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 10]
```diff
@@ -321,14 +228,288 @@ class FederationLookUpPhonebookContactInteractor {
...
+
+    return _ChunkSessionResult(
+      fromMappings: fromMappings,
+      fromThirdParty: fromThirdParty,
+    );
+  }
+
+  /// Queries the identity server for [chunkContacts] unless its URL is already
+  /// covered by [globalSeenThirdPartyUrls]. Returns the matched contacts.
+  Future<Set<Contact>> _processIdentityServerForChunk({
```
Comment (nqhhdev) [inline lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart:397]: This method is only reached when `identitySession != null`, which is only created when `argument.identityServerUrl != null`, so the null-check here is redundant. Could use `identitySession.url` directly instead of re-checking and force-unwrapping `argument.identityServerUrl!`.

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart [hunk 11]
```diff
@@ -55,221 +92,93 @@ class FederationLookUpPhonebookContactInteractor {
...
+                  argument: argument,
+                )
+              : const <Contact>{};
 
           final combinedContacts = chunkContacts.toSet().combineContacts(
-            contactsFromMappings: contactsFromMappings,
-            contactsFromThirdParty: contactsFromThirdParty,
+            contactsFromMappings: {
+              ...sessionsResult.fromMappings,
+              ...contactsFromIdentity,
```
Comment (nqhhdev) [inline lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: Identity server results are spread into contactsFromMappings alongside direct federation mappings. Does combineContacts treat them the same way? Just want to confirm this is intentional and won't cause duplicates or incorrect merge behavior.

**Comment by nqhhdev** [comment]
What do you think if we apply E2E for this feature?

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart (hunk 9 above)
Comment (jbrnzl) [reply lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: Done, I changed convertToHttps directly

File: lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart (hunk 11 above)
Comment (jbrnzl) [reply lib/domain/usecase/contacts/federation_look_up_phonebook_contact_interactor.dart]: Yes, this is intentional. Both sessionsResult.fromMappings and contactsFromIdentity are produced by the same mechanism: a direct lookupMxid call processed through _collectMappings, so they're semantically equivalent and correctly grouped together. combineContacts handles any overlap between the two sources by deduplicating on contact.id and merging via combine(), so no duplicates or incorrect merge behavior will occur. I added a test for that to confirm 👍🏻

**Comment by jbrnzl** [comment]
> What do you think if we apply E2E for this feature?

Good point @nqhhdev. For this change, the risky behaviour (multi-URL federation lookup, identity fallback, merged mappings, dedup) lives in the domain layer and is covered by unit tests on FederationLookUpPhonebookContactInteractor (and related tests).
But I will try to look if I can find an E2E test that would be meaningful. I am not familiar with E2E tests so there is chances I ask you advices on that point.

File: lib/domain/contact_manager/contacts_manager.dart [hunk 12]
```diff
@@ -365,13 +365,15 @@ class ContactsManager {
...
-                      '',
+                  federationUrls:
+                      federationConfigurations.fedServerInformation.baseUrls
+                          ?.map((u) => u.toString())
+                          .toList() ??
+                      [],
+                  identityServerUrl: federationConfigurations
+                      .identityServerInformation
+                      ?.baseUrl
+                      .toString(),
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/domain/contact_manager/contacts_manager.dart:376]: I verified: IdentityServerInformation.baseUrl is a non-nullable Uri, so identityServerInformation?.baseUrl.toString() is already null-safe. Adding ?. before toString() would only trigger invalid_null_aware_operator. No change needed.

**Comment by nqhhdev** [comment]
Hi @jbrnzl could you please add a quick demo for this change?

**Comment by nqhhdev** [comment]
Screenshot 2026-04-28 at 14 37 53

One concern: Identity server queried redundantly for every chunk. Dedup set stores "https://tom.stg.lin-saas.com:443" (from third_party_mappings key tom.stg.lin-saas.com:443), but the check compares against "https://tom.stg.lin-saas.com" (from baseUrl.toString()). Port mismatch → check never hits → TOM gets register+hash_details+lookup on every chunk after chunk 8.

What do you think @jbrnzl ?

**Comment by jbrnzl** [comment]
> Hi @jbrnzl could you please add a quick demo for this change?

Hi @nqhhdev  ! Thank you for reviewing this PR, I really appreciate 👍🏻 
Yes I did not provided visual demos because there is no UI changes, only behavioral changes (more exploration) so more chances to find someone's contact. Do you please have an idea on what exactly I can record to illustrates the changes ?

**Comment by jbrnzl** [comment]
> Screenshot 2026-04-28 at 14 37 53
> One concern: Identity server queried redundantly for every chunk. Dedup set stores "https://tom.stg.lin-saas.com:443" (from third_party_mappings key tom.stg.lin-saas.com:443), but the check compares against "https://tom.stg.lin-saas.com" (from baseUrl.toString()). Port mismatch → check never hits → TOM gets register+hash_details+lookup on every chunk after chunk 8.
> 
> What do you think @jbrnzl ?

Updated 👍🏻

**Comment by nqhhdev** [comment]
> 
> Hi @nqhhdev ! Thank you for reviewing this PR, I really appreciate 👍🏻 Yes I did not provided visual demos because there is no UI changes, only behavioral changes (more exploration) so more chances to find someone's contact. Do you please have an idea on what exactly I can record to illustrates the changes ?

hi @jbrnzl you can follow these docs: https://www.notion.so/linagora/Contact-Syncing-34b87eca754d4614ad9d9a050a1cd956?source=copy_link

**Comment by pm-McFly** [comment]
@nqhhdev The e2ee test sounds interesting but for now I don't see it as a blocker. A unit test that validates we correctly detects all base_urls and verify that the code will process against all is enough.

As well, a demo would have more impact on a feature wise issue, here it is just a behavior aligment to the specs, nothing to really show. As long as the feature still works on our envs where we provide only one base_url, we are good.

The bottom of the issue here was more on the phone number parsing, retrieval and normalization fixed in #3017 
Here it sounds much more of a "BTW I can fix that" ;)

**Review by nqhhdev — APPROVED** [review]
LGTM

**Review by pm-McFly — APPROVED** [review]
